# LAB __ Nivel 0

In [3]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi -c "SELECT 1;"

 ?column? 
----------
        1
(1 row)



# FASE 1 Capa RAW

Lectura del fichero parquet y transformación a fichero CSV (Se necesita importar pandas para poder transformar de parquet a CSV)

In [7]:
import pandas as pd
df = pd.read_parquet('yellow_tripdata_2024-01.parquet')
df.to_csv('yellow_tripdata_2024-01.csv',index=False)

Creación de un nuevo esquema RAW (BRONZE)

In [4]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi <<'SQL'
CREATE SCHEMA IF NOT EXISTS raw;
SQL

CREATE SCHEMA


Creación de la tabla RAW yellow_taxi_raw (Si nos fijamos todas la columnas se quedan en formato texto, no hacemos suposiciones)

In [5]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi <<'SQL'
DROP TABLE IF EXISTS raw.yellow_taxi_raw;

CREATE TABLE raw.yellow_taxi_raw (
    vendor_id TEXT,
    pickup_datetime TEXT,
    dropoff_datetime TEXT,
    passenger_count TEXT,
    trip_distance TEXT,
    ratecode_id TEXT,
    store_and_fwd_flag TEXT,
    pulocation_id TEXT,
    dolocation_id TEXT,
    payment_type TEXT,
    fare_amount TEXT,
    extra TEXT,
    mta_tax TEXT,
    tip_amount TEXT,
    tolls_amount TEXT,
    improvement_surcharge TEXT,
    total_amount TEXT,
    congestion_surcharge TEXT,
    airport_fee TEXT
);
SQL

NOTICE:  table "yellow_taxi_raw" does not exist, skipping


DROP TABLE
CREATE TABLE


Carga CSV a la tabla RAW "yellow.taxi.raw"

In [8]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi <<'SQL'
\copy raw.yellow_taxi_raw FROM 'yellow_tripdata_2024-01.csv' CSV HEADER;
SQL

COPY 2964624


Comprobamos la tabla con datos

In [9]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi <<'SQL'
SELECT COUNT(*) FROM raw.yellow_taxi_raw;
SQL

  count  
---------
 2964624
(1 row)



# FASE 2 - Modelo relacional (CURATED)

Creamos el esquema CURATED (SILVER)

In [10]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi <<'SQL'
CREATE SCHEMA IF NOT EXISTS curated;
SQL

CREATE SCHEMA


Creamos la tabla TRIP 

In [11]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi <<'SQL'
DROP TABLE IF EXISTS curated.trip CASCADE;

CREATE TABLE curated.trip (
    trip_id BIGSERIAL PRIMARY KEY,
    vendor_id INTEGER,
    pickup_datetime TIMESTAMP,
    dropoff_datetime TIMESTAMP,
    passenger_count INTEGER,
    trip_distance NUMERIC,
    fare_amount NUMERIC,
    tip_amount NUMERIC,
    total_amount NUMERIC,
    pulocation_id INTEGER,
    dolocation_id INTEGER
);
SQL

NOTICE:  table "trip" does not exist, skipping


DROP TABLE
CREATE TABLE


Pasamos la información de la capa RAW a la capa CURATED

In [1]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi <<'SQL'
INSERT INTO curated.trip (
    vendor_id,
    pickup_datetime,
    dropoff_datetime,
    passenger_count,
    trip_distance,
    fare_amount,
    tip_amount,
    total_amount,
    pulocation_id,
    dolocation_id
)
SELECT
    vendor_id::INTEGER,
    pickup_datetime::TIMESTAMP,
    dropoff_datetime::TIMESTAMP,
    passenger_count::NUMERIC::INTEGER,
    trip_distance::NUMERIC,
    fare_amount::NUMERIC,
    tip_amount::NUMERIC,
    total_amount::NUMERIC,
    pulocation_id::INTEGER,
    dolocation_id::INTEGER
FROM raw.yellow_taxi_raw;
SQL

INSERT 0 2964624


Comprobamos el volumen de la tabla curated

In [2]:
%%bash
psql postgresql://root:root@localhost:5432/ny_taxi <<'SQL'
SELECT COUNT(*) FROM curated.trip;
SQL

  count  
---------
 2964624
(1 row)



# FASE 3 ANALYTICS -- EDA Indices y Performance (Gold)

In [3]:
%load_ext sql
%sql postgresql://root:root@localhost:5432/ny_taxi

Connecting to 'postgresql://root:***@localhost:5432/ny_taxi'

EDA basico

In [4]:
%%sql
SELECT
    COUNT(*) AS total_trips,
    MIN(pickup_datetime),
    MAX(pickup_datetime)
FROM curated.trip;

Running query in 'postgresql://root:***@localhost:5432/ny_taxi'

1 rows affected.

total_trips,min,max
2964624,2002-12-31 22:59:39,2024-02-01 00:01:15


problema de calidad de datos heredado de la fuente: el csv corresponde a 2024 pero el pickup-datetime es 2002-12-31.

Ingresos por día

In [5]:
%%sql
SELECT
    DATE(pickup_datetime) AS day,
    SUM(total_amount) AS revenue
FROM curated.trip
GROUP BY day
ORDER BY day;

Running query in 'postgresql://root:***@localhost:5432/ny_taxi'

35 rows affected.

day,revenue
2002-12-31,0.0
2009-01-01,127.69
2023-12-31,224.62
2024-01-01,2442843.25
2024-01-02,2282196.02
2024-01-03,2357588.48
2024-01-04,2800511.06
2024-01-05,2728672.52
2024-01-06,2436272.32
2024-01-07,1898102.16


Indice para acelerar la busqueda y filtrado de filas. creamos un indice de pickup_datetime para optimizar consultas analiticas basadas en rangos de tiempo.

In [6]:
%%sql
CREATE INDEX idx_trip_pickup_datetime
ON curated.trip (pickup_datetime);

Running query in 'postgresql://root:***@localhost:5432/ny_taxi'

++
||
++
++

Ahora creamos un explain: 

In [7]:
%%sql
EXPLAIN ANALYZE
SELECT *
FROM curated.trip
WHERE pickup_datetime >= '2024-01-15';

Running query in 'postgresql://root:***@localhost:5432/ny_taxi'

9 rows affected.

QUERY PLAN
Seq Scan on trip (cost=0.00..112674.80 rows=1679196 width=63) (actual time=278.945..517.581 rows=1678072 loops=1)
Filter: (pickup_datetime >= '2024-01-15 00:00:00'::timestamp without time zone)
Rows Removed by Filter: 1286552
Planning Time: 0.326 ms
JIT:
Functions: 2
"Options: Inlining false, Optimization false, Expressions true, Deforming true"
"Timing: Generation 0.316 ms, Inlining 0.000 ms, Optimization 0.396 ms, Emission 3.249 ms, Total 3.961 ms"
Execution Time: 590.036 ms


Lo importante es: cost=0.00..112674.80 rows=1679196 width=63 -- porcentaje de filas devueltas 56% 